tensorflow/core/kernels/scatter_nd_op_gpu.cu.cc

```diff
@@ -72,7 +72,9 @@ template <typename T>
 struct LeftUpdate<std::complex<T>, scatter_nd_op::UpdateOp::SUB> {
   EIGEN_STRONG_INLINE EIGEN_DEVICE_FUNC void operator()(
       std::complex<T>* out, const std::complex<T>& val) {
-    LeftUpdate<std::complex<T>, scatter_nd_op::UpdateOp::ADD>()(out, -val);
+    T* ptr = reinterpret_cast<T*>(out);
+    GpuAtomicAdd(ptr, -val.real());
```

Comment:
Would it be better to use GpuAtomicSub instead?

I'm very puzzled why we seem to be updating the real part twice, here and in ADD. Looks like a bug to me.

Reply:
It just unwrap the function call to `ADD`.

https://github.com/tensorflow/tensorflow/blob/v2.0.0/tensorflow/core/kernels/scatter_nd_op_gpu.cu.cc#L58

https://www.tensorflow.org/api_docs/python/tf/scatter_nd

Reply:
Yes, I see that you are unwrapping the ADD. But if you are unwrapping, why not use GpuAtomicSub instead? And is the ADD implementation actually correct?

Reply:
https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_add
https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_sub

You mean for complex type, should be the real part plus real part, imaginary part plus the imaginary part?

Reply:
That's how complex numbers are added, AFAIK. Thanks.